ddtrace/contrib/celery/signals.py

```diff
@@ -0,0 +1,120 @@
+from ddtrace import Pin
+
+from celery import registry
+
+from . import constants as c
+from .util import (
+    tags_from_context,
+    retrieve_task_id,
+    propagate_span,
+    retrieve_span,
+    remove_span,
+)
+
+
+def trace_prerun(*args, **kwargs):
+    # safe-guard to avoid crashes in case the signals API
+    # changes in Celery
+    task = kwargs.get('sender')
+    task_id = kwargs.get('task_id')
+    if task is None or task_id is None:
+        return
+
+    # retrieve the task Pin or fallback to the global one
+    pin = Pin.get_from(task) or Pin.get_from(task.app)
+    if pin is None:
+        return
+
+    # propagate the `Span` in the current task Context
+    span = pin.tracer.trace(c.WORKER_ROOT_SPAN, service=c.WORKER_SERVICE, resource=task.name)
+    propagate_span(task, task_id, span)
```

Comment:
This part is required to move the span from one signal to another. You can check the weak dictionary implementation for more details. There is also a test that checks for a memory leak. Would be great having your opinion on that.

Sidenote: we can't use objects inside the signal arguments. Any change to dictionary passed there, will alter the Celery behavior.

Reply:
To my understanding (which could be faulty, I'm not well-versed in weak reference usage) it seems as though the only reference to the in-transit span is through the weakref dictionary. If this is the case then aren't the spans subject to being garbage collected at any time? Could this lead to spans being garbage collected before they finish?

Am I missing something?

Reply:
I think this makes sense.

If I'm following things correctly, the tracer object has a strong reference to the span so it won't get garbage collected until the tracer finishes the trace and drops its reference to the span.

In theory having the link to the task as a normal link _should_ be fine, but in the case where there's a memory leak for tasks (e.g. finished tasks being kept around in memory) we'd end up amplifying that problem.

This approach should buy us a bit of safety for fairly minimal risk and complexity.

Reply:
Correct, we have a strong reference in the tracer (`Context` to be accurate). My idea is to avoid increasing the memory leak from our side, even though we can generate a leak from another part of the instrumentation. This change is to propagate tracing data from one signal to the other, reducing at minimum the risk of having a leak in case something goes wrong.

I'm keeping this implementation for now.